src/journal-remote/journal-gatewayd.c

```diff
@@ -885,6 +885,66 @@ static int request_handler_machine(
         return MHD_queue_response(connection, MHD_HTTP_OK, response);
 }
 
+static int request_handler_boots(
+                struct MHD_Connection *connection,
+                void *connection_cls) {
+
+        _cleanup_(MHD_destroy_responsep) struct MHD_Response *response = NULL;
+        RequestMeta *m = ASSERT_PTR(connection_cls);
+        int r;
+        _cleanup_free_ LogId *boots = NULL;
+        size_t n_boots;
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *array = NULL;
+        _cleanup_free_ char *json = NULL;
+
+        assert(connection);
+
+        r = open_journal(m);
+        if (r < 0)
+                return mhd_respondf(connection, r, MHD_HTTP_INTERNAL_SERVER_ERROR, "Failed to open journal: %m");
+
+        r = journal_get_boots(
+                        m->journal,
+                        false,
+                        SIZE_MAX,
+                        &boots, &n_boots);
+
+        if (r < 0)
+                return mhd_respondf(connection, r, MHD_HTTP_INTERNAL_SERVER_ERROR, "Failed to determine boots: %m");
+        if (r == 0)
+                return mhd_respondf(connection, r, MHD_HTTP_INTERNAL_SERVER_ERROR, "No boot found.");
+
+        for (int i = 0; i < (int) n_boots; i++) {
+                char boot_id[SD_ID128_STRING_MAX];
+                sd_id128_to_string(boots[i].id, boot_id);
+
+                r = sd_json_variant_append_arrayb(
+                                &array,
+                                SD_JSON_BUILD_OBJECT(
+                                                SD_JSON_BUILD_PAIR("index", SD_JSON_BUILD_INTEGER(-i)),
+                                                SD_JSON_BUILD_PAIR("boot_id", SD_JSON_BUILD_STRING(boot_id)),
+                                                SD_JSON_BUILD_PAIR("first_entry", SD_JSON_BUILD_UNSIGNED(boots[i].first_usec)),
+                                                SD_JSON_BUILD_PAIR("last_entry", SD_JSON_BUILD_UNSIGNED(boots[i].last_usec))));
+                if (r < 0)
+                        return mhd_respondf(connection, r, MHD_HTTP_INTERNAL_SERVER_ERROR, "Failed to append JSON object: %m");
+        }
+
+        r = sd_json_variant_format(array, 0, &json);
+        if (r < 0) {
+                return mhd_respondf(connection, r, MHD_HTTP_INTERNAL_SERVER_ERROR, "Failed to format JSON data: %m");
+        }
+
+        response = MHD_create_response_from_buffer(r, json, MHD_RESPMEM_MUST_FREE);
```

Comment:
so hmm, there might be quite a few of these, hence we really need to stream this properly, i.e. look for the next bunch whenever the http connection is wrtiable.

i.e. how this works for request_handler_fields() where the actual data is generated by another callback in request_reader_fields()

Reply:
Okay, I first thought I might get away with building the whole array but if it's possible there would be systems with huge list of boots, the overhead of keeping the variant representation and the output buffer would be indeed massive.

I think I mostly see how it should be done when streamed, I'm just not sure what would be the correct way to iterate over the boots. Should the array of boots be kept in the `RequestMeta`? Or should it even not use `journal_get_boots` and iterate in the `MHD_ContentReaderCallback` using `discover_next_id`, only keeping identifiers needed for calling that?